File: p406m/reconstruct_queue.ml
(**
 * @param people: int*int list
 * @return int*int list
 *)
let reconstruct_queue people =
  let sorted = List.stable_sort (fun (h_a,k_a) (h_b,k_b) ->
    if (h_a = h_b) then
      (k_a - k_b) (* increasing k *)
    else
      (h_b - h_a) (* decreasing h *)
  ) people in
  List.fold_left (fun queue (h,k) ->
    (* insert into the position using k value *)
    let rec insert_by_k (h,k) n = function
      | [] -> [(h,k)]
      | hd :: tl as l -> if n = 0 then (h, k) :: l
                         else hd :: insert_by_k (h,k) (n-1) tl in
    insert_by_k (h,k) k queue
  ) [] sorted

(* TEST *)
open Printf
let () =
  let tests = [
    [];
    [(7,0)];
    [(7,0); (4,4); (7,1); (5,0); (6,1); (5,2)]
  ] in
  List.iter (fun test ->
    print_string "After reconstructing queue -> [ ";
    List.iter (fun (h,k) ->
      printf "(%d,%d) " h k
    ) @@ reconstruct_queue @@ test;
    print_endline "]";
  ) tests;
  print_newline ();

File: p406m/reconstruct_queue.js
/**
 * @param {number[][]} people
 * @return {number[][]}
 */
var reconstructQueue = function (people) {
  people.sort((a, b) => {
    if (a[0] === b[0]) {
      return a[1] - b[1]; // increasing k
    } else {
      return b[0] - a[0]; // decreasing h
    }
  });
  let res = [];
  for (let person of people) {
    // insert into the position using k value
    res.splice(person[1], 0, person);
  }
  return res;
};
// TEST
[
  [],
  [[7, 0]],
  [
    [7, 0],
    [4, 4],
    [7, 1],
    [5, 0],
    [6, 1],
    [5, 2],
  ],
].forEach((t) => {
  console.log('After reconstructing queue ->', reconstructQueue(t));
});
